### ISSUE TITLE
☂️ Search: simpler query language inspired by "keyword search"

### ISSUE THREAD

**Issue by jtibshirani** [sourcegraph/sourcegraph#58815]
We plan to bring code search closer to the "keyword search" style that users are used to, where queries are broken up into individual terms that can match broadly across the file name, contents, and symbols. This issue tracks our first round of work on improving the query language, which we aim to merge next release as 'beta'.

**Example query**: `repo:sourcegraph/sourcegraph auth provider`
* `auth provider` matches files that contain both the strings `auth` and `provider` in any order. Before, it only matched files containing the exact string `auth provider`
* There are no changes to filters in this release, so `repo:sourcegraph/sourcegraph` works as usual

### Core changes
- [x] #58799
- [x] Fold prototype into "search keyword" search mode @stefanhengl 
  - [x] Spaces interpreted as ANDs
    - [x] https://github.com/sourcegraph/sourcegraph/pull/59164
  - [x] #59057
  - [x] Support regex patterns using `/.../`
  - [x] https://github.com/sourcegraph/sourcegraph/pull/59686
- [x] Solidify the UX @jtibshirani 
  - https://github.com/sourcegraph/sourcegraph/pull/59707
  - https://github.com/sourcegraph/sourcegraph/pull/59779
  - https://github.com/sourcegraph/sourcegraph/pull/59840
  - https://github.com/sourcegraph/sourcegraph/pull/59915
- [x] https://github.com/sourcegraph/sourcegraph/pull/60047
- [x] Address issue with file match limit, where all top results may be from a single file @keegancsmith

### Usability / performance
- [x] Test that Zoekt can handle AND queries with many clauses @keegancsmith 
- [x] https://github.com/sourcegraph/sourcegraph/issues/59038 @jtibshirani 
- [x] Debug [why new search mode is much slower on S2](https://sourcegraph.slack.com/archives/C05R619V4F8/p1704484790035859) @keegancsmith 
- [x] Promote results with an exact match, so users can still easily copy/ paste in snippets @keegancsmith 
  - https://github.com/sourcegraph/zoekt/pull/728
  - https://github.com/sourcegraph/sourcegraph/pull/59940

### Testing + polish
- [x] https://github.com/sourcegraph/docs/pull/55
- [x] Telemetry @jtibshirani 
  - https://github.com/sourcegraph/sourcegraph/pull/59097
  - https://github.com/sourcegraph/sourcegraph/pull/59993
  - https://github.com/sourcegraph/sourcegraph/pull/60022
- [x] Complete "Keyword search quality plan" ([separate doc](https://docs.google.com/document/d/1skfewTKAcGR0oZrSGYLLBtFhlAHo3ORH5jQQAo2pcNk/edit))

/cc @sourcegraph/search-platform

**Comment by keegancsmith** [comment]
### DONE Test that Zoekt can handle AND queries with many clauses

TL;DR from random real world queries I tried I got acceptable performance. For s2 I normally got around 1s when ANDing, but with literal it would be much faster (0.2s). This is slower, but acceptable.

I got started on writing something which would try out a bunch of queries and collect data and graph it. But I ended up just writing the part which sampled out real strings from a codebase. When trying out a few samples at different token counts I got pretty consistent performance. In particular I did a bunch of testing with 2, 10, 15 and 20 token lengths.

My mental model of why this worked out fine is the larger the number of tokens, the less documents to consider. So even though it is querying the index more, it balances out so we have pretty consistent performance.

What I did notice is for a few of the large AND queries the ranking was terrible. I think it will be quite important to fix that (ie promote atom counts on same line, etc). In particular if you have an atom that is a common word, you quickly run into limits since each atom matching counts towards your limit. So I often ran into a top ranked document which was just a large file that happened to contain all the terms as substrings.

In the interest of someone maybe wanting to take this further see below for the code which outputs strings from the sourcegraph codebase.

<details>

```python
#!/usr/bin/env python3

import collections
import subprocess
import random

def gen_corpus():
    "return a corpus of strings from go code"
    proc = subprocess.run(
        args=('rg', '-t', 'go', '-o', '--no-filename', '"[^"]+?"'),
        cwd='/Users/keegan/src/github.com/sourcegraph/sourcegraph/',
        capture_output=True,
        check=True,
    )
    for line in proc.stdout.split(b'\n'):
        yield(line.decode('utf-8'))

literals = collections.defaultdict(set)
for s in gen_corpus():
    if ' ' not in s:
        continue
    if s.startswith('" ') or s.endswith(' "'):
        continue
    literals[len(s.split())].add(s)

def summarize(upto=20):
    for count, ss in sorted(literals.items()):
        if count > upto:
            break
        s = repr(ss)
        if len(s) > 80:
            s = s[:80]
        print(count, len(ss), s)

def sample(tokens, count=10):
    # remove items which contain keywords
    keywords = set(['and', 'not', 'or'])
    population = [s for s in literals[tokens] if keywords.isdisjoint(s.lower().split())]

    count = min(count, len(population))
    for s in random.sample(population, count):
        try:
            print(eval(s))
            print(' AND '.join(eval(s).split()))
            print()
        except SyntaxError:
            pass

#summarize()
sample(20)
```

</details>

**Comment by jtibshirani** [comment]
> Make sure we can quantify improvements to search experience before/ after rollout

I looked into our metrics collection for searches and think we are already tracking the right metrics:
* We log `SearchSubmitted` and `SearchResultClicked`, so we can see how many searches resulted in clicks vs. not
* Thanks to @rrhyne we also report a more nuanced 'search success' metric which captures whether a user clicked then followed up with an action like copying code, searching history, etc.
* We also log if search throws an error (`SearchResultsFetchFailed`) and whether it returns any results (`SearchResultsNonEmpty`)

All logged events contain the feature flags that were enabled, so we can compare metrics from before and after the flag is enabled for an instance. We need to be aware of confounding factors, like the fact that they also upgraded to a new version which could affect "search success".

One complexity is that users will be able to toggle the new search syntax from the UI. This is controlled through the search pattern type, which is not stored as part of the event. I think this is okay, we'll still get good signal at the instance level as to whether enabling the feature was helpful. It'd be good to add tracking for how often users disable the toggle though.

**Comment by jtibshirani** [comment]
5.3 release status ✅ (on track). Feature work completed, now moving on to testing and squashing bugs.

Link to test plan: https://docs.google.com/document/d/1skfewTKAcGR0oZrSGYLLBtFhlAHo3ORH5jQQAo2pcNk/edit
